Feat/supabase add third party

clairefro opened the pull request:
*Note: for local dev must:
1.  update Supabase redirect host url to a localhost tunnel
--- start ngrok: `ngrok http 8910 -host-header="localhost:8910"`
--- in Supabase dash: Auth > Settings > Site Url: `<ngrok url>/supabase/welcome`
1. add `SUPABASE_JWT_SECRET` (dash > Settings > API) in a **local** .env file. 

Adds responsive third party auth button container for auth methods with third party auth options. 

Currently only used in Supabase, but can be shared with others like Firebase. 

![image](https://user-images.githubusercontent.com/9841162/115166988-a29b5b00-a06a-11eb-9e23-c37907151d7c.png)

~TODO: Fix Github/Google integration errors.~ 

~[Github]~
~flashes `Unable to exchange external code:` in url on redirect. Auth does not complete~
~[Google]~
~On successful auth: `Unhandled Rejection (AuthError): Unexpected error in authentication.: Hash does not contain state`~

dthyresson commented:
@clairefro Excellent start and I like the container with the provider options.

As you already know, but writing here for others, the issue was that this Supabase project is an early project and as such the auth api version it used was a few revisions out of date and it has not been updated (it didn't have scopes etc).

So, this should be back ready for testing, but there are few issues:

* GitHib needs `scopes` and a possible `redirectTo` like:

```js
          await logIn({
            provider: 'github',
            scopes: 'user repo',
            redirectTo: 'https://somewhereover.ngrok.io/supabase',
          })
```

* Does it not need `await`?
* Unfortunately, the callback redirect when the access_token codes back after its code has been exchanged uses params what the Redwood Azure provider grabs a hold of first and it tries to use the token and thus the "hash" error.
* I tried to remove the provider altogether and while the error no longer happens, Supabase didn't store the token/user info because then Netlify Identity tries to grab it since they are both GoTrue based.

I'm not sure how to handle the Azure/Netlify issue except to figure out a way they can all live together nicely in one import.

It may mean that the nice auth indicator on the HomePage can't be there since that loads all of the providers.

* Also, in your local env, make sure that your `SUPABASE_JWT_SECRET` is set. You can get that from the Supabase Admin Dashboard > Settings > API

dthyresson commented:
@clairefro I just had an idea to fix the "other providers" grab the access token issue.

What if supabase redirected to a "welcome" or some landing page (ie, not the supabase "Provider" page and not home) and that page *only* has the Supabase auth provider initialized.

Both HomePage and ProviderPage import and init all the clients. If there was a callback page for just Supabase that said "Thanks!" which stores the sessions and then a button link or a redirect back to /supabase I bet that could work.

Worth a try!

The thing to avoid is importing and initing all the providers.

clairefro commented:
> What if supabase redirected to a "welcome" or some landing page (ie, not the supabase "Provider" page and not home) and that page _only_ has the Supabase auth provider initialized.

Cool idea - I set the redirect to `/supabase/welcome` and the hash error went away for both Github and Google!

Now I get new `ERR_SSL_PROTOCOL_ERROR`s, which I suspect are ngrok related. deep diving 🔍  
![image](https://user-images.githubusercontent.com/9841162/115327966-b5825e00-a144-11eb-99bb-b0e253f4432c.png)

clairefro commented:
Working now for Github and Google after adding `AuthProvider` with supabase client to the `SupabaseWelcomePage`

![image](https://user-images.githubusercontent.com/9841162/115329508-45290c00-a147-11eb-9769-66ce305c8ae3.png)

![image](https://user-images.githubusercontent.com/9841162/115329567-540fbe80-a147-11eb-89d7-e9960a7394de.png)

I will add a 3 second redirect back to Supabase auth from welcome page

clairefro commented:
Have a hacky version working. for Supabase Github/Google. No scope was needed for Github

Do we want `<AuthResults>` or `<PollCurrentVersionCell>` - they seem to provide similar data

![Peek 2021-04-19 21-08](https://user-images.githubusercontent.com/9841162/115336225-9d660b00-a153-11eb-8e84-b70ddeccb89b.gif)

dthyresson commented:
> Have a hacky version working for Supabase Github/Google. No scope was needed for Github

Doesn't look hacky at all -- that's exactly how I envisioned it working. Kudos for 🎉 use.
 
> Do we want `<AuthResults>` or `<PollCurrentVersionCell>` - they seem to provide similar data

I think we'll want to refine the presentation of both of these components in another issue/pr.

The intent of `AuthResults` is to show the user info and some basic checks that the content is sensible (ie, it has an email if needed and other attributes.

The intent of `PollCurrentVersionCell` is to make sure that the token refreshes after it expires -- it the expires at should refresh after say 1 hour in most cases (but depends of provider).

I'll create a new issue to discuss.

dthyresson reviewed (commented):
Looks great!

Had a few comments but nothing major.

dthyresson commented:
@clairefro Seems like you can remove the Draft and open up for others to review, but looks great!

clairefro commented on .env.defaults:
```diff
@@ -31,4 +31,4 @@ MAGICLINK_PUBLIC="pk_test_3471FA4515E86534"
 SUPABASE_KEY="eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoiYW5vbiIsImlhdCI6MTU5OTQ4OTc4MCwiZXhwIjoxOTE1MDY1NzgwfQ.lBsGLQsITqu-eSt-bWNOgemmXWPp49EEhbHLFGpRF8M"
 SUPABASE_URL="https://ojkjaqetdyfiibiurmlc.supabase.co"
-
+SUPABASE_JWT_SECRET="d7b7302f-4549-4a95-831b-aed62f8aba28"
```
oopsie :) I saw other secrets out in the open there and thought it was OK 

Thanks for addressing - I will update PR tonight

clairefro commented:
working late tonight - will get to this tomorrow

clairefro commented on web/src/components/ThirdPartyProviderButton/ThirdPartyProviderButton.js:
```diff
@@ -0,0 +1,31 @@
+        src={logo}
+        style={{
+          maxWidth: '25px',
+          filter: disabled ? 'grayscale(1)' : 'grayscale(0)',
```
yessir! done

clairefro commented:
- added loading spinner to third party auth container 
- removed `SUPABASE_JWT_SECRET` from plainsight 
- moved `PollCurrentVersionCell` to be adjacent to `AuthResults` - see how we like this

![Peek 2021-04-22 15-33](https://user-images.githubusercontent.com/9841162/115793143-793b4180-a380-11eb-974b-068d7a1d1d82.gif)

Tobbe commented:
Maybe I'm just slow, but the first time I saw that redirect page I didn't have time to read it all before it was gone. I suggest increasing the timer to 5, or even 10 seconds. And provide a button to manually go back for those who are too impatient to wait.

clairefro commented:
> Maybe I'm just slow, but the first time I saw that redirect page I didn't have time to read it all before it was gone. I suggest increasing the timer to 5, or even 10 seconds. And provide a button to manually go back for those who are too impatient to wait.

I've now upped the timeout to 5 seconds. Clicking the link will take you back earlier

dthyresson commented on web/src/components/Spinner/Spinner.js:
```diff
@@ -0,0 +1,6 @@
+import 'react-loader-spinner/dist/loader/css/react-spinner-loader.css'
+import Loader from 'react-loader-spinner'
+
+export default () => {
```
Instead of adding a new component, import and css, did you consider using:

https://tailwindcss.com/docs/animation#spin

dthyresson commented on web/src/components/ThirdPartyProviderContainer/ThirdPartyProviderContainer.js:
```diff
@@ -0,0 +1,26 @@
+export default ({ providers, onProviderClick, loading, ...rest }) => {
+  if (loading) {
+    return (
+      <div className="flex justify-center items-center">
```
Instead of a spinner component, could you use:

https://tailwindcss.com/docs/animation#spin

?

dthyresson reviewed (commented):
Just wondering if spinner couldn't be replaced with Tailwind svg animations instead of importing more css and and another component lib.

morganmspencer commented:
> Just wondering if spinner couldn't be replaced with Tailwind svg animations instead of importing more css and and another component lib.

Could it be a Heroicon (already installed) with the spinner TW class on it?

clairefro commented:
Great suggestions guys

how about a spinning `view-grid` icon from `@heroicon`
![Peek 2021-04-22 20-30](https://user-images.githubusercontent.com/9841162/115814370-c089f800-a3a9-11eb-93b6-c18e5ced1843.gif)

dthyresson commented:
@clairefro Tailwind spinner works for me. We can also swap out the svg or other icon later, but think it's time to merge.

Thanks and am going to :shipit: !
